chips/apollo3/src/chip.rs

```diff
@@ -3,31 +3,89 @@
 use core::fmt::Write;
 use cortexm4;
 use kernel::Chip;
+use kernel::InterruptService;
 
-use crate::ble;
-use crate::gpio;
-use crate::iom;
-use crate::nvic;
-use crate::stimer;
-use crate::uart;
-
-pub struct Apollo3 {
+pub struct Apollo3<I: InterruptService + 'static> {
     mpu: cortexm4::mpu::MPU,
     userspace_kernel_boundary: cortexm4::syscall::SysCall,
     scheduler_timer: cortexm4::systick::SysTick,
+    interrupt_service: &'static I,
 }
 
-impl Apollo3 {
-    pub unsafe fn new() -> Apollo3 {
-        Apollo3 {
+impl<I: InterruptService + 'static> Apollo3<I> {
+    pub unsafe fn new(interrupt_service: &'static I) -> Self {
```

Comment:
I think `apollo3_driver_definitions` would need to be passed into the chip, as the chip might want internally use a driver which has been overridden by the board. Otherwise we would have a mismatch of board using one  driver implementation, and the board another.

In this case the macro below, probably could just be a function, something like `apollo3_default_drivers_get` or something similar.

Reply:
I think that if a chip needs to use a driver internally, that driver should be a field on the `Chip` struct itself, revealing that the Chip depends on that specific driver (and thereby showing that driver is not optional). If we want that driver to further be replaceable by multiple implementations, then introducing a trait for that particular driver would be appropriate. But I think we want to make it explicitly not possible for the chip to depend on the particular contents of whatever struct implements `InterruptService`, and instead depend only on the interface exposed by `InterruptService`.

Reply:
It seems like having that driver be an argument to `chip::new()` is reasonable to handle this case.